In [95]:
import pandas as pd
import numpy as np
import os

import matplotlib.pyplot as plt

In [96]:
#we need to cycle through our data so that we can then combine our data to then save it all as a single 
# dataset.

In [97]:
os.getcwd()

'/Users/Alex/Alex/IU_Masters_Program/DSCI-4492_Data_Viz/Final Project'

In [98]:
path = '2010-2022 poverty data (no 2020)/Data'

file_list = os.listdir('2010-2022 poverty data (no 2020)/Data')

In [99]:
column_count = 0
for i in file_list:
    df = pd.read_csv(path + '/' + i)
    print(f'{i}: {df.shape}\n')
    column_count += df.shape[1]
    
print(column_count)

2021.csv: (842, 375)

2022.csv: (849, 375)

2019.csv: (841, 369)

2018.csv: (839, 369)

2015.csv: (831, 369)

2014.csv: (829, 279)

2016.csv: (832, 369)

2017.csv: (838, 369)

2013.csv: (829, 279)

2012.csv: (826, 279)

2010.csv: (819, 279)

2011.csv: (823, 279)

3990


In [100]:
for i in file_list:
    print(f"df_{i[:4]},")

df_2021,
df_2022,
df_2019,
df_2018,
df_2015,
df_2014,
df_2016,
df_2017,
df_2013,
df_2012,
df_2010,
df_2011,


In [101]:
df_2021 = pd.read_csv(path + '/' + file_list[0])
df_2022 = pd.read_csv(path + '/' + file_list[1])
df_2019 = pd.read_csv(path + '/' + file_list[2])
df_2018 = pd.read_csv(path + '/' + file_list[3])
df_2015 = pd.read_csv(path + '/' + file_list[4])
df_2014 = pd.read_csv(path + '/' + file_list[5])
df_2016 = pd.read_csv(path + '/' + file_list[6])
df_2017 = pd.read_csv(path + '/' + file_list[7])
df_2013 = pd.read_csv(path + '/' + file_list[8])
df_2012 = pd.read_csv(path + '/' + file_list[9])
df_2010 = pd.read_csv(path + '/' + file_list[10])
df_2011 = pd.read_csv(path + '/' + file_list[11])

In [102]:
df_list = [df_2021,df_2022,df_2019,
           df_2018,df_2015,df_2014,
           df_2016,df_2017,df_2013,
           df_2012,df_2010,df_2011]
#df_list

In [103]:
df_2021.loc[0]

GEO_ID                                                    Geography
NAME                                           Geographic Area Name
S1701_C01_001E    Estimate!!Total!!Population for whom poverty s...
S1701_C01_001M    Margin of Error!!Total!!Population for whom po...
S1701_C01_002E    Estimate!!Total!!Population for whom poverty s...
                                        ...                        
S1701_C03_061E    Estimate!!Percent below poverty level!!UNRELAT...
S1701_C03_061M    Margin of Error!!Percent below poverty level!!...
S1701_C03_062E    Estimate!!Percent below poverty level!!UNRELAT...
S1701_C03_062M    Margin of Error!!Percent below poverty level!!...
Unnamed: 374                                                    NaN
Name: 0, Length: 375, dtype: object

In [104]:
col_list = ['datasets']
for i in df_list:
    for j in i.columns:
        col_list.append(i[j][0])

In [105]:
len(col_list)

3991

In [106]:
col_list

['datasets',
 'Geography',
 'Geographic Area Name',
 'Estimate!!Total!!Population for whom poverty status is determined',
 'Margin of Error!!Total!!Population for whom poverty status is determined',
 'Estimate!!Total!!Population for whom poverty status is determined!!AGE!!Under 18 years',
 'Margin of Error!!Total!!Population for whom poverty status is determined!!AGE!!Under 18 years',
 'Estimate!!Total!!Population for whom poverty status is determined!!AGE!!Under 18 years!!Under 5 years',
 'Margin of Error!!Total!!Population for whom poverty status is determined!!AGE!!Under 18 years!!Under 5 years',
 'Estimate!!Total!!Population for whom poverty status is determined!!AGE!!Under 18 years!!5 to 17 years',
 'Margin of Error!!Total!!Population for whom poverty status is determined!!AGE!!Under 18 years!!5 to 17 years',
 'Estimate!!Total!!Population for whom poverty status is determined!!AGE!!Under 18 years!!Related children of householder under 18 years',
 'Margin of Error!!Total!!Populatio

In [113]:
col_list_unique = []

for i in col_list:
    if i in col_list_unique:
        pass
    else:
        col_list_unique.append(i)
    
#col_list_unique = list(col_set)

len(col_list_unique)

1128

In [115]:
#col_list_unique

In [116]:
col_list_unique.index("datasets")

0

In [117]:
existing_columns = pd.DataFrame(columns = col_list_unique)
existing_columns

Empty DataFrame
Columns: [datasets, Geography, Geographic Area Name, Estimate!!Total!!Population for whom poverty status is determined, Margin of Error!!Total!!Population for whom poverty status is determined, Estimate!!Total!!Population for whom poverty status is determined!!AGE!!Under 18 years, Margin of Error!!Total!!Population for whom poverty status is determined!!AGE!!Under 18 years, Estimate!!Total!!Population for whom poverty status is determined!!AGE!!Under 18 years!!Under 5 years, Margin of Error!!Total!!Population for whom poverty status is determined!!AGE!!Under 18 years!!Under 5 years, Estimate!!Total!!Population for whom poverty status is determined!!AGE!!Under 18 years!!5 to 17 years, Margin of Error!!Total!!Population for whom poverty status is determined!!AGE!!Under 18 years!!5 to 17 years, Estimate!!Total!!Population for whom poverty status is determined!!AGE!!Under 18 years!!Related children of householder under 18 years, Margin of Error!!Total!!Population for whom poverty status is determined!!AGE!!Under 18 years!!Related children of householder under 18 years, Estimate!!Total!!Population for whom poverty status is determined!!AGE!!18 to 64 years, Margin of Error!!Total!!Population for whom poverty status is determined!!AGE!!18 to 64 years, Estimate!!Total!!Population for whom poverty status is determined!!AGE!!18 to 64 years!!18 to 34 years, Margin of Error!!Total!!Population for whom poverty status is determined!!AGE!!18 to 64 years!!18 to 34 years, Estimate!!Total!!Population for whom poverty status is determined!!AGE!!18 to 64 years!!35 to 64 years, Margin of Error!!Total!!Population for whom poverty status is determined!!AGE!!18 to 64 years!!35 to 64 years, Estimate!!Total!!Population for whom poverty status is determined!!AGE!!60 years and over, Margin of Error!!Total!!Population for whom poverty status is determined!!AGE!!60 years and over, Estimate!!Total!!Population for whom poverty status is determined!!AGE!!65 years and over, Margin of Error!!Total!!Population for whom poverty status is determined!!AGE!!65 years and over, Estimate!!Total!!Population for whom poverty status is determined!!SEX!!Male, Margin of Error!!Total!!Population for whom poverty status is determined!!SEX!!Male, Estimate!!Total!!Population for whom poverty status is determined!!SEX!!Female, Margin of Error!!Total!!Population for whom poverty status is determined!!SEX!!Female, Estimate!!Total!!Population for whom poverty status is determined!!RACE AND HISPANIC OR LATINO ORIGIN!!White alone, Margin of Error!!Total!!Population for whom poverty status is determined!!RACE AND HISPANIC OR LATINO ORIGIN!!White alone, Estimate!!Total!!Population for whom poverty status is determined!!RACE AND HISPANIC OR LATINO ORIGIN!!Black or African American alone, Margin of Error!!Total!!Population for whom poverty status is determined!!RACE AND HISPANIC OR LATINO ORIGIN!!Black or African American alone, Estimate!!Total!!Population for whom poverty status is determined!!RACE AND HISPANIC OR LATINO ORIGIN!!American Indian and Alaska Native alone, Margin of Error!!Total!!Population for whom poverty status is determined!!RACE AND HISPANIC OR LATINO ORIGIN!!American Indian and Alaska Native alone, Estimate!!Total!!Population for whom poverty status is determined!!RACE AND HISPANIC OR LATINO ORIGIN!!Asian alone, Margin of Error!!Total!!Population for whom poverty status is determined!!RACE AND HISPANIC OR LATINO ORIGIN!!Asian alone, Estimate!!Total!!Population for whom poverty status is determined!!RACE AND HISPANIC OR LATINO ORIGIN!!Native Hawaiian and Other Pacific Islander alone, Margin of Error!!Total!!Population for whom poverty status is determined!!RACE AND HISPANIC OR LATINO ORIGIN!!Native Hawaiian and Other Pacific Islander alone, Estimate!!Total!!Population for whom poverty status is determined!!RACE AND HISPANIC OR LATINO ORIGIN!!Some other race alone, Margin of Error!!Total!!Population for whom poverty status is determined!!RACE AND HISPANI

What the above tells me is that there are 1128 columns that are shared between the 10 datasets, when we ideally would have 375, which is the most columns in a single dataset.  

As such, we'll need to look up each specific column for each dataset that we want to use.

See the "README.md" file in "Working Dataset" for a list of the types of columns we kept.

# Finalized Dataset

Here is the basic analysis of our finalized dataset comprised of the 10 years of data.

First, we look at the data to make sure the datasets joined correctly.

In [143]:
os.getcwd()

'/Users/Alex/Alex/IU_Masters_Program/DSCI-4492_Data_Viz/Final Project'

In [144]:
path = 'Working Dataset'

df = pd.read_csv(path + '/2010_2022_pov_data.csv')

/var/folders/z1/ly8kccd50r56yj1_qq1np6q80000gp/T/ipykernel_99046/1907957291.py:3: DtypeWarning: Columns (26,27,28,29,30,31,44,45,46,47,48,49,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,198,199,202,203,204,205,208,209,210,211,214,215,216,217,220,221,222,223,232,233,234,235,238,239,240,241,244,245,246,247,248,249,256,257,258,259,262,263,264,265,268,269,270,271,273,274,275,276,277,296,297,302,303,308,309,346,347,348,349,350,351) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(path + '/2010_2022_pov_data.csv')


In [145]:
df

GEO_ID                                  NAME  S1701_C01_001E  \
0     0500000US01003               Baldwin County, Alabama          180686   
1     0500000US01015               Calhoun County, Alabama          116553   
2     0500000US01043               Cullman County, Alabama           79692   
3     0500000US01049                DeKalb County, Alabama           70453   
4     0500000US01051                Elmore County, Alabama           72069   
...              ...                                   ...             ...   
9981  0500000US72113          Ponce Municipio, Puerto Rico          128156   
9982  0500000US72127       San Juan Municipio, Puerto Rico          330936   
9983  0500000US72135       Toa Alta Municipio, Puerto Rico           65885   
9984  0500000US72137       Toa Baja Municipio, Puerto Rico           72613   
9985  0500000US72139  Trujillo Alto Municipio, Puerto Rico           66604   

      S1701_C01_001M  S1701_C02_001E  S1701_C02_001M  S1701_C03_001E  \
0                985           23610            4573            13.1   
1               1987           29082            4105            25.0   
2                834           16848            3811            21.1   
3                682           13788            4362            19.6   
4               3261            7923            2292            11.0   
...              ...             ...             ...             ...   
9981             849           68118            4856            53.2   
9982            1171          126728            8503            38.3   
9983              28           18491            3232            28.1   
9984              11           24484            3586            33.7   
9985              11           20202            4606            30.3   

      S1701_C03_001M  S1701_C01_002E  S1701_C01_002M  ...  S1701_C03_061E  \
0                2.5           41865             592  ...             NaN   
1                3.4           26847             296  ...             NaN   
2                4.8           18892             588  ...             NaN   
3                6.2           18233             460  ...             NaN   
4                3.1           18189             306  ...             NaN   
...              ...             ...             ...  ...             ...   
9981             3.8           22543              34  ...            72.1   
9982             2.6           50952              89  ...            63.9   
9983             4.9           10890               2  ...            59.4   
9984             4.9           11665               4  ...            71.1   
9985             6.9           10557               2  ...            76.4   

      S1701_C03_061M  Unnamed: 368  S1701_C01_062E  S1701_C01_062M  \
0                NaN           NaN             NaN             NaN   
1                NaN           NaN             NaN             NaN   
2                NaN           NaN             NaN             NaN   
3                NaN           NaN             NaN             NaN   
4                NaN           NaN             NaN             NaN   
...              ...           ...             ...             ...   
9981             6.0           NaN        127633.0           841.0   
9982             4.2           NaN        328313.0          1292.0   
9983            12.9           NaN         65722.0            31.0   
9984             7.5           NaN         72448.0            15.0   
9985             9.5           NaN         66055.0           130.0   

      S1701_C02_062E  S1701_C02_062M  S1701_C03_062E  S1701_C03_062M  \
0                NaN             NaN             NaN             NaN   
1                NaN             NaN             NaN             NaN   
2                NaN             NaN             NaN             NaN   
3                NaN             NaN             NaN             NaN   
4                NaN             NaN             NaN             NaN   
...           

In [146]:
columns_to_keep = ['NAME', 
                   'S1701_C01_001E', 'S1701_C02_001E', 'S1701_C01_002E', 
                   'S1701_C02_002E', 'S1701_C01_004E', 'S1701_C02_004E', 
                   'S1701_C01_005E', 'S1701_C02_005E', 'S1701_C01_006E', 
                   'S1701_C02_006E', 'S1701_C01_007E', 'S1701_C02_007E']

In [149]:
df = df[columns_to_keep]

In [154]:
columns_renamed = {'S1701_C01_001E': 'total_pop',
                   'S1701_C02_001E': 'below_pov',
                   'S1701_C01_002E': 'total_under_18',
                   'S1701_C02_002E': 'below_pov_under_18',
                   'S1701_C01_004E': 'total_18_to_64',
                   'S1701_C02_004E': 'below_pov_18_to_64',
                   'S1701_C01_005E': 'total_over_64',
                   'S1701_C02_005E': 'below_pov_over_64',
                   'S1701_C01_006E': 'total_male',
                   'S1701_C02_006E': 'below_pov_male',
                   'S1701_C01_007E': 'total_female',
                   'S1701_C02_007E': 'below_pov_female'}

In [156]:
df = df.rename(columns = columns_renamed)

In [157]:
df[df['NAME'] == 'San Juan Municipio, Puerto Rico']

NAME  total_pop  below_pov  total_under_18  \
814   San Juan Municipio, Puerto Rico     386027     151075           84043   
1636  San Juan Municipio, Puerto Rico     384662     156275           82390   
2461  San Juan Municipio, Puerto Rico     376313     150338           79606   
3289  San Juan Municipio, Puerto Rico     368673     155142           76365   
4117  San Juan Municipio, Puerto Rico     360203     158007           72726   
4947  San Juan Municipio, Puerto Rico     350863     160738           69187   
5778  San Juan Municipio, Puerto Rico     340979     136122           65321   
6615  San Juan Municipio, Puerto Rico     332876     132263           61725   
7453  San Juan Municipio, Puerto Rico     315765     125457           55394   
8293  San Juan Municipio, Puerto Rico     312498     124927           53202   
9134  San Juan Municipio, Puerto Rico     332086     132480           52971   
9982  San Juan Municipio, Puerto Rico     330936     126728           50952   

      below_pov_under_18  total_18_to_64  below_pov_18_to_64 total_over_64  \
814                47678          238574               84813         63410   
1636               46996          235975               88801         66297   
2461               45787          229145               83602         67562   
3289               47246          224115               87046         68193   
4117               46400          217473               87284         70004   
4947               46564           51811               33195         69114   
5778               37490           48956               27151         64952   
6615               34655           46589               25083         61492   
7453               32585           42497               24445         55287   
8293               30306           41152               23369         53014   
9134               32117           41586               25919         52866   
9982               27908           40247               21790         50713   

     below_pov_over_64  total_male  below_pov_male  total_female  \
814              18584      177263           63688        208764   
1636             20478      176134           67830        208528   
2461             20949      172455           64663        203858   
3289             20850      168540           65165        200133   
4117             24323      165740           68688        194463   
4947             46491      211923           87601         78471   
5778             37121      204262           76848         74715   
6615             34422      198509           73994         71803   
7453             32478      189089           71834         67948   
8293             30118      186033           69001         67002   
9134             32012      196995           73170         73270   
9982             27669      195357           68149         72694   

      below_pov_female  
814              87387  
1636             88445  
2461             85675  
3289             89977  
4117             89319  
4947             35561  
5778             33967  
6615             31425  
7453             31626  
8293             30361  
9134             30206  
9982             29650

We need to include the years from each dataset that these values were pulled from.

But let's take a look at the % variance for each county to see how far off they are from each other.

In [171]:
#df = pd.DataFrame({'GROUP': [1, 2, 1, 2, 1], 'VALUE': [5, 2, 10, 20, 7]})
result = df.groupby('NAME')['total_pop'].agg(['max','min','mean'])
result['diff'] = (result['max']-result['min'])/result['mean']
print(result['diff'].info())

print('Max percentage difference: '.format(result['diff'].max()))
print('Min percentage difference: '.format(min(result['diff'])))
print('Mean percentage difference: '.format(np.mean(result['diff'])))

<class 'pandas.core.series.Series'>
Index: 869 entries, Ada County, Idaho to Yuma County, Arizona
Series name: diff
Non-Null Count  Dtype  
--------------  -----  
869 non-null    float64
dtypes: float64(1)
memory usage: 13.6+ KB
None
Max percentage difference: 
Min percentage difference: 
Mean percentage difference: 


In [172]:
max(result['diff'])

0.5718960776230285

In [173]:
min(result['diff'])

0.0

In [174]:
np.mean(result['diff'])

0.10747467659787514

In [175]:
np.median(result['diff'])

0.08143660298879568